src/libsync/clientsideencryption.cpp

```diff
@@ -0,0 +1,502 @@
+#include "clientsideencryption.h"
+#include "account.h"
+#include "capabilities.h"
+#include "networkjobs.h"
+
+#include <openssl/rsa.h>
+#include <openssl/evp.h>
+#include <openssl/pem.h>
+#include <openssl/err.h>
+
+#include <map>
+#include <string>
+
+#include <cstdio>
+
+#include <QDebug>
+#include <QLoggingCategory>
+#include <QFileInfo>
+#include <QDir>
+#include <QJsonObject>
+
+#include "wordlist.h"
+
+namespace OCC
+{
+
+Q_LOGGING_CATEGORY(lcCse, "sync.clientsideencryption", QtInfoMsg)
+
+QString baseUrl = QStringLiteral("ocs/v2.php/apps/client_side_encryption/api/v1/");
+QString baseDirectory = QDir::homePath() + QStringLiteral("/.nextcloud-keys/");
+
+namespace {
+    void handleErrors(void)
+    {
+        qCInfo(lcCse()) << "Error handling encryption or decryption";
+        ERR_print_errors_fp(stdout); // This line is not printing anything.
+        fflush(stdout);
+    }
+
+    int encrypt(unsigned char *plaintext,
+                int plaintext_len,
+                unsigned char *aad,
+                int aad_len,
+                unsigned char *key,
+                unsigned char *iv,
+                unsigned char *ciphertext,
+                unsigned char *tag)
+    {
+        EVP_CIPHER_CTX *ctx;
+        int len;
+        int ciphertext_len;
+
+        /* Create and initialise the context */
+        if(!(ctx = EVP_CIPHER_CTX_new())) {
+            handleErrors();
+        }
+
+        /* Initialise the encryption operation. */
+        if(1 != EVP_EncryptInit_ex(ctx, EVP_aes_256_gcm(), NULL, NULL, NULL)) {
+            handleErrors();
+        }
+
+        /* Set IV length if default 12 bytes (96 bits) is not appropriate */
+        if(1 != EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, 16, NULL)) {
```

Comment:
Why use a non-standard nonce size?

Reply:
Code exerpt taken from the openssl examples, What should be an standard nonce size?